modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,660 @@
+    public static final long DFLT_FILE_MAX_SIZE = 32 * 1024 * 1024 * 1024L;
+
+    /** Default off heap buffer size in bytes. */
+    public static final int DFLT_BUFFER_SIZE = 32 * 1024 * 1024;
+
+    /** Default minimal batch size to flush in bytes. */
+    public static final int DFLT_FLUSH_SIZE = 8 * 1024 * 1024;
+
+    /** Directory to store performance statistics files. Placed under Ignite work directory. */
+    public static final String PERFORMANCE_STAT_DIR = "performanceStatistics";
+
+    /** Factory to provide I/O interface. */
+    private final FileIOFactory fileIoFactory = new RandomAccessFileIOFactory();
+
+    /** Performance statistics enabled flag. */
+    private volatile boolean enabled;
+
+    /** Performance statistics file writer worker. */
+    @Nullable private volatile FileWriter fileWriter;
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** @param ctx Kernal context. */
+    public FilePerformanceStatisticsWriter(GridKernalContext ctx) {
+        log = ctx.log(getClass());
+
+        this.ctx = ctx;
+    }
+
+    /** @return {@code True} if collecting performance statistics enabled. */
+    public boolean performanceStatisticsEnabled() {
+        return enabled;
+    }
+
+    /** Starts collecting performance statistics. */
+    public synchronized void start() {
+        if (enabled)
+            return;
+
+        FileWriter writer = fileWriter;
+
+        // Writer is stopping.
+        if (writer != null) {
+            try {
+                writer.shutdown().get();
+            }
+            catch (IgniteCheckedException e) {
+                throw new IgniteException("Failed to wait for previous writer stopping.", e);
+            }
+        }
+
+        assert fileWriter == null;
+
+        try {
+            File file = statisticsFile(ctx);
+
+            U.delete(file);
+
+            FileIO fileIo = fileIoFactory.create(file);
+
+            fileIo.position(0);
+
+            fileWriter = new FileWriter(ctx, fileIo, DFLT_FILE_MAX_SIZE, DFLT_BUFFER_SIZE, DFLT_FLUSH_SIZE, log);
+
+            new IgniteThread(fileWriter).start();
+
+            enabled = true;
+
+            log.info("Performance statistics writer started [file=" + file.getAbsolutePath() + ']');
+        }
+        catch (IOException | IgniteCheckedException e) {
+            log.error("Failed to start performance statistics writer.", e);
+
+            throw new IgniteException("Failed to start performance statistics writer.", e);
+        }
+    }
+
+    /** Stops collecting performance statistics. */
+    public IgniteInternalFuture<Void> stop() {
+        synchronized (this) {
+            if (!enabled)
+                return new GridFinishedFuture<>();
+
+            enabled = false;
+        }
+
+        log.info("Stopping performance statistics writer.");
+
+        FileWriter fileWriter = this.fileWriter;
+
+        if (fileWriter != null)
```

Comment:
Can we rewrite flow a bit to make this code more readable.
```
if (fileWriter == null)
    return new GridFinishedFuture<>();

return fileWriter.shutdown();
```

Reply:
Done.